### PR TITLE
Tensor pr

### PR DESCRIPTION
**Pull request by Akilesh2**
5 tensor Augmentations to OpenVX

### CODE REVIEW

**Comment by fiona-gladwin** [comment]
@Akilesh2 can you please remove all the files in amd_nn, Those files are not supposed to be changed

File: amd_openvx_extensions/amd_rpp/source/image/copy.cpp
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/image/copy.cpp:1]: The other function names in this file must also be changed right?
Can you please check the lk/rocalTensor branch : [link](https://github.com/LakshmiKumar23/MIVisionX/blob/lk/rocalTensor/amd_openvx_extensions/amd_rpp/source/CopybatchPD.cpp)
Also the name of the file has been changed

File: amd_openvx_extensions/amd_rpp/source/image/nop.cpp
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/image/nop.cpp:1]: Please check this file too [link](https://github.com/LakshmiKumar23/MIVisionX/blob/lk/rocalTensor/amd_openvx_extensions/amd_rpp/source/NopbatchPD.cpp)

Change name of the file

File: amd_openvx_extensions/amd_rpp/source/image/kernel_rpp.cpp [hunk 1]
```diff
@@ -1876,6 +1906,140 @@ VX_API_CALL vx_node VX_API_CALL  vxExtrppNode_SequenceRearrange(vx_graph graph,v
...
 }
 
+//tensor
+
+VX_API_ENTRY vx_node VX_API_CALL vxExtrppNode_Brightness(vx_graph graph, vx_tensor pSrc, vx_tensor srcROI, vx_tensor pDst, vx_array alpha, vx_array beta, vx_scalar inputLayout, vx_scalar outputLayout, vx_scalar roiType) {
+    vx_node node = NULL;
+    vx_context context = vxGetContext((vx_reference)graph);
+    if (vxGetStatus((vx_reference)context) == VX_SUCCESS) {
+        vx_uint32 dev_type = getGraphAffinity(graph);
+        vx_scalar DEV_TYPE = vxCreateScalar(vxGetContext((vx_reference)graph), VX_TYPE_UINT32, &dev_type);
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/image/kernel_rpp.cpp:1916]: DEV_TYPE should not be in capital letters please change it to devType
Same comment applicable in all the newly introduced API
Do not touch the old ones

File: amd_openvx_extensions/amd_rpp/source/image/kernel_rpp.cpp [hunk 2]
```diff
@@ -1876,6 +1906,140 @@ VX_API_CALL vx_node VX_API_CALL  vxExtrppNode_SequenceRearrange(vx_graph graph,v
...
+        vx_reference params[] = {
+            (vx_reference)pSrc,
+            (vx_reference)srcROI,
+            (vx_reference)pDst,
+            (vx_reference)dst_width,
+            (vx_reference)dst_height,
+            (vx_reference)interpolation_type,
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/image/kernel_rpp.cpp:1966]: Use camel case variable names for the above 3, dst_width, dst_height and interpolation_type

File: rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py [hunk 3]
```diff
@@ -318,4 +318,4 @@ def draw_patches(img,idx, bboxes):
     image = image.transpose([1,2,0])
     image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR )
     image = cv2.UMat(image).get()
-    cv2.imwrite(str(idx)+"_"+"train"+".png", image)
+    cv2.imwrite(str(idx)+"_"+"train"+".png", image)
```
Comment (fiona-gladwin) [inline rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:321]: Revert this change

File: utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp [hunk 4]
```diff
@@ -760,7 +760,6 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
                   << "rocalSSDRandomCrop" << std::endl;
         image1 = rocalSSDRandomCrop(handle, input1, true);
     }
-    break;
```
Comment (fiona-gladwin) [inline utilities/rocAL/rocAL_unittests/rocAL_unittests.cpp:763]: Why this change here?

File: amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+#include "internal_publishKernels.h"
+
+struct CopyLocalData {
+    vxRppHandle handle;
+    Rpp32u device_type;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    size_t tensor_size;
+    vx_enum in_tensor_type;
+    vx_enum out_tensor_type;
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp:32]: Change variable names to camel case to maintain uniformity

File: amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+static vx_status VX_CALLBACK refreshCopy(vx_node node, const vx_reference *parameters, vx_uint32 num, CopyLocalData *data) {
+    vx_status status = VX_SUCCESS;
+    if (data->device_type == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_HIP
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HIP, &data->pSrc, sizeof(data->pSrc)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_HIP, &data->pDst, sizeof(data->pDst)));
+#endif
+    } else if (data->device_type == AGO_TARGET_AFFINITY_CPU) {
+            STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HOST, &data->pSrc, sizeof(data->pSrc)));
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp:43]: Decrease indent in these 2 lines here

File: amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include "internal_publishKernels.h"
+
+struct NopLocalData {
+    vxRppHandle handle;
+    Rpp32u device_type;
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp:27]: Use camel case here

File: amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    if (data->device_type == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_HIP
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HIP, &data->pSrc, sizeof(data->pSrc)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_HIP, &data->pDst, sizeof(data->pDst)));
+#endif
+    } else if (data->device_type == AGO_TARGET_AFFINITY_CPU) {
+        vx_enum in_tensor_type = vx_type_e::VX_TYPE_UINT8;
+        vx_enum out_tensor_type = vx_type_e::VX_TYPE_UINT8;
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_DATA_TYPE, &in_tensor_type, sizeof(in_tensor_type)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_DATA_TYPE, &out_tensor_type, sizeof(out_tensor_type)));
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp:43]:
```suggestion
        vx_enum in_tensor_type = vx_type_e::VX_TYPE_UINT8;
        vx_enum out_tensor_type = vx_type_e::VX_TYPE_UINT8;
        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_DATA_TYPE, &in_tensor_type, sizeof(in_tensor_type)));
        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_DATA_TYPE, &out_tensor_type, sizeof(out_tensor_type)));
```

Are these 4 lines required?

File: amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    void *roi_tensor_ptr;
+    RpptROI * roi_ptr;
+    RpptRoiType roiType;
+    Rpp32s input_layout;
+    Rpp32s output_layout;
+    size_t inputTensorDims[VX_TENSOR_DIMS];
+    size_t outputTensorDims[VX_TENSOR_DIMS];
+    vx_enum in_tensor_type;
+    vx_enum out_tensor_type;
+    RpptImagePatch *dstImgSize;
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp:45]: Please use camel case only for all variables

**Review by fiona-gladwin — CHANGES_REQUESTED** [review]
@Akilesh2 Please check the comments and make appropriate changes

**Comment by fiona-gladwin** [comment]
> 5 tensor Augmentations to OpenVX

Please follow a practice of giving a good description for the PR
Explaining the file changes and addition

File: amd_openvx_extensions/amd_nn/src/activation_layer.cpp [hunk 10]
```diff
@@ -106,7 +106,7 @@ static vx_status VX_CALLBACK initializeActivationLayer(vx_node node, const vx_re
 {
     ActivationLayerLocalData * data = new ActivationLayerLocalData;
     memset(data, 0, sizeof(*data));
-    ERROR_CHECK_STATUS(createGraphHandle(node, &data->handle));
+    ERROR_CHECK_STATUS(createRPPHandle(node, &data->handle));
```
Comment (sampath1117) [inline amd_openvx_extensions/amd_nn/src/activation_layer.cpp:109]: why RPP handle has been changed here?
it is under amd_nn

@Akilesh2 we already made this change in performance PR and faced lot of issues
Please do not change files other than the RPP / relevelant files

File: amd_openvx_extensions/amd_nn/src/activation_layer.cpp [hunk 11]
```diff
@@ -171,7 +171,7 @@ static vx_status VX_CALLBACK uninitializeActivationLayer(vx_node node, const vx_
     ERROR_CHECK_MIOPEN_STATUS(miopenDestroyTensorDescriptor(data->inputDescriptor));
     ERROR_CHECK_MIOPEN_STATUS(miopenDestroyTensorDescriptor(data->outputDescriptor));
     if (data) {
-        ERROR_CHECK_STATUS(releaseGraphHandle(node, data->handle));
+        ERROR_CHECK_STATUS(releaseRPPHandle(node, data->handle));
```
Comment (sampath1117) [inline amd_openvx_extensions/amd_nn/src/activation_layer.cpp:174]: Same comment as before
This file should not be changed

File: amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp [hunk 12]
```diff
@@ -135,7 +135,7 @@ static vx_status VX_CALLBACK initializeBatchNormalizationLayer(vx_node node, con
 {
     BatchNormLayerLocalData * data = new BatchNormLayerLocalData;
     memset(data, 0, sizeof(*data));
-    ERROR_CHECK_STATUS(createGraphHandle(node, &data->handle));
+    ERROR_CHECK_STATUS(createRPPHandle(node, &data->handle));
```
Comment (sampath1117) [inline amd_openvx_extensions/amd_nn/src/batch_normalization_layer.cpp:138]: All files under amd_nn should not be changed
Applicable to all such instances

**Comment by fiona-gladwin** [comment]
> @Akilesh2 can you please remove all the files in amd_nn, Those files are not supposed to be changed

@Akilesh2 This comment is not addressed yet

**Comment by Akilesh2** [comment]
@fiona-gladwin  I have resolved the PR comments, can you please verify it.

File: amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 2, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxFinalizeKernel(kernel));
+    }
+    if (status != VX_SUCCESS) {
+    exit:
+        vxRemoveKernel(kernel);
+        return VX_FAILURE;
+    }
+    return status;
+}
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp:180]: Add line at EOF

File: amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 2, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxFinalizeKernel(kernel));
+    }
+    if (status != VX_SUCCESS) {
+    exit:
+        vxRemoveKernel(kernel);
+        return VX_FAILURE;
+    }
+    return status;
+}
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp:158]: Add line at EOF

File: rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py (hunk 3 above)
Comment (fiona-gladwin) [reply rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:321]: You need to add a new line at EOF to resolve this

**Comment by sampath1117** [comment]
@fiona-gladwin @Akilesh2 
There are 98 file changes in this PR, i can see most of them are due to files restructuring / renaming

If you guys did not had a plan to break this PR into smaller PR's, please think of it
@fiona-gladwin i am assuming you have run the rocAL unittests, performance tests in both C++ and python in this PR branch
if not @Akilesh2 can you please share the docker and commands to run the tests in the PR branch
we have to do the complete testing

**Review by shobana-mcw — COMMENTED** [review]
Please address the review comments. Some comments needs to be updated to similar files. Please let me know if you have any doubts.

File: amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+
+static vx_status VX_CALLBACK processBrightness(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    RppStatus rpp_status = RPP_SUCCESS;
+    vx_status return_status = VX_SUCCESS;
+    BrightnessLocalData *data = NULL;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_HIP
+        refreshBrightness(node, parameters, num, data);
+        rpp_status = rppt_brightness_gpu(data->pSrc, data->srcDescPtr, data->pDst, data->dstDescPtr,  data->alpha, data->beta, data->roiPtr, data->roiType, data->handle->rppHandle);
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp:124]: For tensor we do not have the OCL support in RPP

[the remark this answers is not in the document]
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp:124]: But Rajy said we need to add OCL support ryt? I remember having some discussion with Rajy.

**Comment by fiona-gladwin** [comment]
@Akilesh2 are all comments in this PR addressed?

**Comment by Akilesh2** [comment]
@fiona-gladwin  I have addressed all the PR comments

File: amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h [hunk 16]
```diff
@@ -118,9 +118,16 @@ vx_status WarpAffinebatchPD_Register(vx_context);
 vx_status WarpPerspectivebatchPD_Register(vx_context);
 vx_status Resizetensor_Register(vx_context);
 
+//tensor
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h:121]: @Akilesh2 this comment not resolved

File: amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+
+#include "internal_publishKernels.h"
+
+struct CopyLocalData {
+    vxRppHandle handle;
+    Rpp32u deviceType;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    size_t tensor_size;
+    vx_enum inputTensorType;
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp:31]: All these varibales needs to be in camelCase please make the change, Keep it uniform

File: amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+
+static vx_status VX_CALLBACK refreshCopy(vx_node node, const vx_reference *parameters, vx_uint32 num, CopyLocalData *data) {
+    vx_status status = VX_SUCCESS;
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_HIP
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HIP, &data->pSrc, sizeof(data->pSrc)));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_HIP, &data->pDst, sizeof(data->pDst)));
+#endif
+    } else if (data->deviceType == AGO_TARGET_AFFINITY_CPU) {
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HOST, &data->pSrc, sizeof(data->pSrc)));
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp:43]: Remove one indentation here

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp [hunk 19]
```diff
@@ -1876,6 +1875,170 @@ VX_API_CALL vx_node VX_API_CALL  vxExtrppNode_SequenceRearrange(vx_graph graph,v
...
+    }
+    return node;
+}
+
+VX_API_ENTRY vx_node VX_API_CALL vxExtrppNode_Copy(vx_graph graph, vx_tensor pSrc, vx_tensor pDst) {
+    vx_node node = NULL;
+    vx_context context = vxGetContext((vx_reference)graph);
+    if (vxGetStatus((vx_reference)context) == VX_SUCCESS) {
+        vx_uint32 dev_type = getGraphAffinity(graph);
+        vx_scalar DEV_TYPE = vxCreateScalar(vxGetContext((vx_reference)graph), VX_TYPE_UINT32, &dev_type);
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:1906]: All the variables DEV_TYPE should be changed to devType please make the change

File: amd_openvx_extensions/amd_rpp/CMakeLists.txt [hunk 20]
```diff
@@ -46,90 +46,95 @@ include_directories(${Boost_INCLUDE_DIRS})
...
+        source/image/TensorMatrixMultiply.cpp
+        source/image/TensorMultiply.cpp
+        source/image/TensorSubtract.cpp
+        source/image/ThresholdingbatchPD.cpp
+        source/image/VignettebatchPD.cpp
+        source/image/WarpAffinebatchPD.cpp
+        source/image/WarpPerspectivebatchPD.cpp
+        source/image/kernel_rpp.cpp
+        source/image/internal_publishKernels.cpp
+        source/tensor/Brightness.cpp
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/CMakeLists.txt:131]: @Akilesh2
Please make the change atleast for tensor aug

File: amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp (hunk 9 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp:40]: @Akilesh2 This comment is not addressed

File: amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    vx_float32 *alpha;
+    vx_float32 *beta;
+    RpptDescPtr srcDescPtr;
+    RpptDesc srcDesc;
+    RpptDesc dstDesc;
+    RpptDescPtr dstDescPtr;
+    void *roiTensorPtr;
+    RpptROI *roiPtr;
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp:34]: @Akilesh2 did you check if this is feasible?

File: amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+    vx_float32 *beta;
+    RpptDescPtr srcDescPtr;
+    RpptDescPtr dstDescPtr;
+    RpptROI *roiPtr;
+    RpptRoiType roiType;
+    Rpp32s inputLayout;
+    Rpp32s outputLayout;
+    size_t inputTensorDims[VX_TENSOR_DIMS];
+    size_t ouputTensorDims[VX_TENSOR_DIMS];
+};
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp:40]: The scope of this roiTensorPtr can also be reduced I feel
Please check the feasibility and check in one of the tensor branch after this change if it works as expected then make changes to all the files

File: amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+    RpptDescPtr dstDescPtr;
+    void *roiTensorPtr;
+    RpptROI *roiPtr;
+    RpptRoiType roiType;
+    Rpp32s inputLayout;
+    Rpp32s outputLayout;
+    size_t inputTensorDims[VX_TENSOR_DIMS];
+    size_t ouputTensorDims[VX_TENSOR_DIMS];
+    vx_enum inputTensorType;
+    vx_enum outputTensorType;
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp]: @Akilesh2 why this is not addressed?

File: amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+static vx_status VX_CALLBACK initializeCropMirrorNormalize(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    CropMirrorNormalizeLocalData *data = new CropMirrorNormalizeLocalData;
+    memset(data, 0, sizeof(*data));
+#if ENABLE_OPENCL
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU)
+        THROW("initialize : CMN, OpenCL backend is not supported")
+#endif
+    int roi_type;
```
[the remark this answers is not in the document]
Comment (sampath1117) [reply amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp:152]: @Akilesh2 
This comment has not been addressed
Please make the change

File: amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (Akilesh2) [reply amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp]: i have already addressed this comment.

File: amd_openvx_extensions/amd_rpp/CMakeLists.txt [hunk 25]
```diff
@@ -46,90 +46,95 @@ include_directories(${Boost_INCLUDE_DIRS})
...
+        source/image/TensorLookup.cpp
+        source/image/TensorMatrixMultiply.cpp
+        source/image/TensorMultiply.cpp
+        source/image/TensorSubtract.cpp
+        source/image/ThresholdingbatchPD.cpp
+        source/image/VignettebatchPD.cpp
+        source/image/WarpAffinebatchPD.cpp
+        source/image/WarpPerspectivebatchPD.cpp
+        source/image/kernel_rpp.cpp
+        source/image/internal_publishKernels.cpp
```
[the remark this answers is not in the document]
Comment (Akilesh2) [reply amd_openvx_extensions/amd_rpp/CMakeLists.txt]: Done

File: amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp (hunk 17 above)
Comment (Akilesh2) [reply amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp:31]: Done

File: amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp (hunk 18 above)
Comment (Akilesh2) [reply amd_openvx_extensions/amd_rpp/source/tensor/Copy.cpp:43]: Done

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp (hunk 19 above)
Comment (Akilesh2) [reply amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp]: done

File: amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp [hunk 26]
```diff
@@ -0,0 +1,258 @@
...
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    vx_float32 *multiplier;
+    vx_float32 *offset;
+    vx_uint32 *mirror;
+    RpptDescPtr srcDescPtr;
+    RpptDescPtr dstDescPtr;
+    RpptROI *roiPtr;
+    RpptRoiType roiType;
+    Rpp32s inputLayout;
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp:37]: Layout as local

File: amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+
+static vx_status VX_CALLBACK processNop(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    NopLocalData *data = NULL;
+
+    vx_status status = VX_SUCCESS;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_HIP
+        refreshNop(node, parameters, num, data);
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp:77]: Why we are checking device? If we need to just refresh we need not have this if condition iself

File: amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+
+static vx_status VX_CALLBACK initializeNop(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    NopLocalData *data = new NopLocalData;
+    memset(data, 0, sizeof(*data));
+#if ENABLE_HIP
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_ATTRIBUTE_AMD_HIP_STREAM, &data->handle.hipstream, sizeof(data->handle.hipstream)));
+#endif
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[2], &data->deviceType, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    refreshNop(node, parameters, num, data);
+
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp:93]: Remove this empty line

File: amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+    Rpp32u deviceType;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    vx_uint32 *resizeHeight;
+    vx_uint32 *resizeWidth;
+    RpptDescPtr srcDescPtr;
+    RpptDescPtr dstDescPtr;
+    RpptROI *roiPtr;
+    RpptRoiType roiType;
+    Rpp32s inputLayout;
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp:36]: layout to be local

File: amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp [hunk 30]
```diff
@@ -0,0 +1,264 @@
...
+    ResizeLocalData *data = NULL;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    refreshResize(node, parameters, num, data);
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_HIP
+        rpp_status = rppt_resize_gpu(data->pSrc, data->srcDescPtr, data->pDst, data->dstDescPtr, data->dstImgSize, (RpptInterpolationType)data->interpolationType, data->roiPtr, data->roiType, data->handle->rppHandle);
+        return_status = (rpp_status == RPP_SUCCESS) ? VX_SUCCESS : VX_FAILURE;
+#endif
+    }
+    if (data->deviceType == AGO_TARGET_AFFINITY_CPU) {
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp:131]: This should be else if

File: amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp [hunk 31]
```diff
@@ -0,0 +1,264 @@
...
+static vx_status VX_CALLBACK processResize(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    RppStatus rpp_status = RPP_SUCCESS;
+    vx_status return_status = VX_SUCCESS;
+    ResizeLocalData *data = NULL;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    refreshResize(node, parameters, num, data);
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_HIP
+        rpp_status = rppt_resize_gpu(data->pSrc, data->srcDescPtr, data->pDst, data->dstDescPtr, data->dstImgSize, (RpptInterpolationType)data->interpolationType, data->roiPtr, data->roiType, data->handle->rppHandle);
+        return_status = (rpp_status == RPP_SUCCESS) ? VX_SUCCESS : VX_FAILURE;
```
Comment (sampath1117) [inline amd_openvx_extensions/amd_rpp/source/tensor/Resize.cpp:128]: I am just wondering if it is better to do a hipStreamSynchronize() at the end of each RPP HIP call
Are we doing a synchronize() anywhere else in rocAL?
@fiona-gladwin @shobana-mcw 
Please share your thoughts on this

File: amd_openvx_extensions/amd_rpp/CMakeLists.txt [hunk 32]
```diff
@@ -46,90 +46,95 @@ include_directories(${Boost_INCLUDE_DIRS})
...
+        source/image/TensorMatrixMultiply.cpp
+        source/image/TensorMultiply.cpp
+        source/image/TensorSubtract.cpp
+        source/image/ThresholdingbatchPD.cpp
+        source/image/VignettebatchPD.cpp
+        source/image/WarpAffinebatchPD.cpp
+        source/image/WarpPerspectivebatchPD.cpp
         source/internal_publishKernels.cpp
+        source/kernel_rpp.cpp
+        source/tensor/Brightness.cpp
```
Comment (fiona-gladwin) [inline amd_openvx_extensions/amd_rpp/CMakeLists.txt:133]: Move these tensor includes before 
source/internal_publishKernels.cpp

**Review by sampath1117 — COMMENTED** [review]
@Akilesh2 
Please address review comments

**Comment by fiona-gladwin** [comment]
@Akilesh2 Please remove all the file changes under rocal_pybind folder from this PR
Add that as a separate PR

File: amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp [hunk 33]
```diff
@@ -0,0 +1,147 @@
...
+    return status;
+}
+
+static vx_status VX_CALLBACK initializeNop(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    NopLocalData *data = new NopLocalData;
+    memset(data, 0, sizeof(*data));
+#if ENABLE_HIP
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_ATTRIBUTE_AMD_HIP_STREAM, &data->handle.hipstream, sizeof(data->handle.hipstream)));
+#endif
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[2], &data->deviceType, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
```
Comment (shobana-mcw) [inline amd_openvx_extensions/amd_rpp/source/tensor/Nop.cpp:83]: Do we need to initialize or refresh?

File: rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py [hunk 34]
```diff
@@ -105,7 +105,7 @@ def main():
         elif augmentation_name == "fish_eye":
             output = fn.fish_eye(images)
         elif augmentation_name == "vignette":
-            output = fn.vignette(images)
+            output = fn.vignette(images, vignette=50.0)
```
Comment (shobana-mcw) [inline rocAL/rocAL_pybind/examples/rocAL_api_python_unittest.py:108]: Why is this change part of this PR?

File: rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py [hunk 35]
```diff
@@ -26,11 +26,14 @@
...
+import cupy as cp
 
-def draw_patches(img,idx):
+def draw_patches(image,idx,device_type):
     #image is expected as a tensor, bboxes as numpy
     import cv2
-    image = img.transpose([0,1,2])
+    if device_type == "gpu":
+        image= cp.asnumpy(image)
+    # image = img.transpose([0,1,2])
```
Comment (shobana-mcw) [inline rocAL/rocAL_pybind/examples/rocAL_api_tf_classification_reader.py:36]: remove commented line.

**Comment by fiona-gladwin** [comment]
Merged
